---
title: "Problem Set 3. Due Thurs March 2 5pm"
author: "Shanshan Pei"
date: "Feb 23, 2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, comment = "#>")
```


# Submission

1. Fork this repository to your own account
1. Make changes to the `README.Rmd` file (including the author field above).
1. Knit the file to HTML format and add it to your git repository (`git add`)
1. Submit a pull request with your Rmd and knitted HTML files.

# Overview 

You will examine a [time-course of mRNA expression][1] of barley powdery mildew fungus in immunocompromised plants (*Arabidopsis*). The fungus Blumeria graminis f. sp. hordei (called *Bgh* from now on)

Counts were generated using the [`maSigPro` software][3] and deposited in [NCBI GEO][2].

[1]: http://www.ncbi.nlm.nih.gov/pubmed/23175786
[2]: http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE43163  
[3]: http://bioinformatics.oxfordjournals.org/content/30/18/2598.full

Some important identifiers for the data set:

| Name | Description |
| ---- | ----------- |
| `A6` | *Bgh* isolate expressing other AVRA effectors |
| `K1` | *Bgh* isolate expressing the cognate AVRA1 effector for MLA1|
| `pps` | *Arabidopsis* plants |
| `B12` | *Arabidopsis* plants expressing MLA1-HA|

We will download tables from GEO containing counts of mRNA abundance from both the fungal sample.

# Raw data

First we need load the data.

```{r load_raw_data, message = FALSE}
library(tidyverse)
# bgh data
bgh_url <- 'http://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE43163&format=file&file=GSE43163_CompleteCountTable_Bgh.txt.gz'

# 2 info and blank lines at top, skip them
raw_data <- read_tsv(bgh_url, skip = 2)

# the header for the first column is "NA", set it manually
names(raw_data)[1] <- 'gene.name'

raw_data <- as_data_frame(raw_data)
```

# Problems


## Problem 1

Now that the raw data are loaded, your first task is to tidy the data with `tidyr`. The `cleaned_data` should look something like this:

```
# A tibble: 310,896 Ã— 6
             gene.name plant fungus time.point   rep value
*                <chr> <chr>  <chr>      <chr> <chr> <int>
1             bgh04079   B12     A6       6hpi     1    13
2             bgh01634   B12     A6       6hpi     1    31
3  bghG000012000001001   B12     A6       6hpi     1   121
4  bghG000012000002001   B12     A6       6hpi     1     3
5             bgh00757   B12     A6       6hpi     1   253
6             bgh01273   B12     A6       6hpi     1    45
7             bgh01274   B12     A6       6hpi     1    32
8             bgh01277   B12     A6       6hpi     1     9
9             bgh06140   B12     A6       6hpi     1    47
10            bgh05774   B12     A6       6hpi     1    42
# ... with 310,886 more rows
```

The key steps are to `gather()` the data into key-value pairs and then `separate()` the information from the key column into new columns.

```{r problem_1}
# add tidying code here
raw_data %>% 
  gather(key, value, -gene.name) %>% 
  separate(key, into = c('plant', 'fungus','time.point','rep'), sep = '_') -> cleaned_data
# report the cleaned_data by just naming it, uncomment the following line:
cleaned_data
```

## Problem 2

You need to translate what you learned on the command line into the R world.

Which plant has the highest expression of any gene in the 6hpi time point?

```{bash, eval=FALSE}
raw_data | awk '$4 == "6hpi"' | sort -k6nr | head -n 1 | cut -f2
```

```{r}
cleaned_data %>% 
  filter(time.point == "6hpi") %>%
  filter(value == max(value)) %>% 
  select(plant)
```

Which plant / fungus pair has the highest expression in the 18hpi time point?

```{bash, eval=FALSE}
raw_data | awk '$4 == "18hpi"' | sort -k6nr | head -n 1 | cut -f2,3
```

```{r}
cleaned_data %>% 
  filter(time.point == "18hpi") %>%
  filter(value == max(value)) %>% 
  select(plant,fungus)
```

## Problem 3

Identify the top 3 most consistently differentially expressed genes between the earliest and latest time points for each combination of `plant` and `fungus` strains.

- "Differential expression"" is the difference between `value` (i.e., gene expression level) between time points.

- "Consistency" is the smallest variance in `value` between replicates.

### Strategy

1. Create a new table from the cleaned data by moving each `hpi` value to a new column name, with counts for each in the column (hint: use a `tidyr` verb). It is helpful to reformat the `hpi` values by converting from e.g. `6hpi` to `hpi.6`. You can use `mutate` to do this, i.e.:

```{r problem_3}
library(stringr)
# replace the 6hpi format with the hpi.6 format and generate a new column containing just time value.
cleaned_data %>% 
  mutate(time.point = str_replace(time.point, 'hpi', ''), time.point = str_c('hpi.', time.point)) -> cleaned_data_1
# generate new columns that contain expression values for each different time point
cleaned_data_1 %>% 
  spread(time.point, value) ->cleaned_data_2 
```

2. Create a new column containing the expression difference between the relevant time points.
```{r add_expr_diff_column}
cleaned_data_2 %>% mutate(expr.diff = hpi.24 - hpi.6) -> expr_data
```

3. Calculate summary statistics (mean and variance) of the expression differences by grouping (hint) the `gene.name`, `plant`, and `virus` columns.
```{r expr_summary_stats}
expr_data %>% 
  group_by(plant, fungus, gene.name) %>%
  summarise(expr.diff.mean = mean(expr.diff), expr.diff.var = var(expr.diff)) -> expr_summary_stats
```

4. Sort by these statistics and use the `dplyr` verb `slice` to pull the ones you want (i.e., the top 3). Note you will have to remove `gene.name` from the grouping so that sorting works.
```{r sort_stats}
expr_summary_stats %>%
  ungroup()%>%
  arrange(plant, fungus, desc(abs(expr.diff.mean)), expr.diff.var) %>%
  group_by(plant, fungus) %>%
  slice(1:3) -> top_genes

top_genes
```


## Problem 4

Now examine the above final data frame above and write a few sentences putting inline code in least 3 places. For example, There are `r nrow(mtcars)` rows of data in `mtcars`.
```{r include=FALSE}
top_genes %>% ungroup()%>%select(gene.name) %>% distinct() %>% as_vector() %>% paste(collapse=' | ')->unique_genes
```

The final data frame `top_genes` has `r nrow(top_genes)` rows and `r ncol(top_genes)` columns of data and contains following unique genes: `r unique_genes`.


## Problem 5

Plot the expression (`value`) by time (`hpi`) for the above genes. Format the plot as follows:

1. Plot each point.
1. Connect the points with a smooth line 
1. Create a facet for each `plant` / `bgh` pair 
1. Assign a different color to each gene 
1. Assign a different shape to each replicate 
1. Add a descriptive title to the plot 
1. Use the "black & white" theme

Extra credit: add error bars to the plot (use `geom_errorbar`).

```{r problem_5}
library(ggplot2)
# generate a vector containing unique gene names
top_genes %>% ungroup()%>%select(gene.name) %>% distinct() %>% as_vector() -> unique_gene_list

# filter data
cleaned_data_1 %>% 
  filter(gene.name %in% unique_gene_list) %>%
  mutate(time.point = str_replace(time.point, 'hpi.', '')) -> filtered_data

# generate graph
ggplot(filtered_data, aes(x = factor(time.point, levels = c("6","12","18","24")), color= gene.name, shape = rep, y=value, group = gene.name)) + geom_point() + geom_line() + facet_wrap(plant ~fungus) + ggtitle('MOLB7621 Problem Set 3 - Q5') + xlab('Time point') + ylab('Expression') + theme_bw()

# generate a new graph with error bars

cleaned_data_1 %>% filter(gene.name %in% unique_gene_list) %>% mutate(time.point = str_replace(time.point, 'hpi.', '')) %>% group_by(plant, fungus, gene.name, time.point) %>% summarise(mean=mean(value), sd=sd(value)) -> data_stat

ggplot(data_stat, aes(x = factor(time.point, levels = c("6","12","18","24")), color= gene.name, y=mean, group=gene.name)) + geom_point() + geom_line() + facet_wrap(plant ~fungus) + ggtitle('MOLB7621 Problem Set 3 - Q5 with error bars') + xlab('Time point') + ylab('Expression') + theme_bw() + geom_errorbar(aes(ymin = mean-sd, ymax = mean+sd))

```
